### ISSUE TITLE
Bff session store db handle peak loads

### ISSUE THREAD

**Issue by erkanmaras** [DuendeSoftware/Support#518]
**Which version of Duende BFF are you using?**
BFF 2.0

**Which version of .NET are you using?**
.Net 6

**Describe the bug**

Hi ✋ , We are using BFF framework with the serverside sessions feature and sessions are stored via EF session store to Azure-Sql server.

Our website is hosted on the azure app service. After every deployment swap, the BFF database is reaching its limits(especially CPU) because we have thousands of signalR connections, and they are trying to reconnect after deployment. But many of them fail. We tried the randomizing signalR connection reconnection time, but it didn't effect much.

The question is, **how can we manage these types of peak loads on the BFF session store?** 

We plan to implement a custom **IUserSessionStore** with in-memory cache support, and right before the deployment swap, we'll populate the cache from the session-store database. Do you think this makes sense?

Do you have any advice regarding this situation?

Thanks ...

**Comment by brockallen** [comment]
Hi. I'm a little confused on what's happening here. You are saying that something in your SignalR connection connects to the user session store? I feel like I'm missing part of the picture.

**Comment by erkanmaras** [comment]
> Hi. I'm a little confused on what's happening here. You are saying that something in your SignalR connection connects to the user session store? I feel like I'm missing part of the picture.

Actually, after the azure app service deployment swap, thousand of signalR connections trying to reconnect at the almost same time, and this cause a lot of select queries on the UserSessions table for a couple of minutes, so we are looking for a solution to these peak loads.

**Comment by brockallen** [comment]
> Actually, after the azure app service deployment swap, thousand of signalR connections trying to reconnect at the almost same time, and this cause a lot of select queries on the UserSessions table for a couple of minutes, so we are looking for a solution to these peak loads.

Well... that's just a normal scale issue, right? In other words, I don't have a silver bullet for you, sorry. At this point it's just normal web and DB "stuff". If our default EF store implementation doesn't really work for your scenario, you can always implement the store interface yourself and tune it for your needs.

**Comment by erkanmaras** [comment]
Yeah, I just wanted to make sure we are not missing something, we'll implement a custom session storage.
But an official Redis session storage would be nice especially high load scenarios.

**Comment by brockallen** [comment]
Turns out a redis implementation is harder than you'd think. You can have a look at what was started here https://github.com/DuendeSoftware/BFF/pull/31 and perhaps use it as a starting point or even rekindle the PR if you can come up with a good solution.
